config-application-package/src/main/java/com/yahoo/config/model/application/provider/FilesApplicationPackage.java

```diff
@@ -613,12 +611,17 @@ private void preprocessXML(File destination, File inputXml, Zone zone) throws IO
 
     @Override
     public ApplicationPackage preprocess(Zone zone, DeployLogger logger) throws IOException {
-        IOUtils.recursiveDeleteDir(preprocessedDir);
-        IOUtils.copyDirectory(appDir, preprocessedDir, -1,
-                              (__, name) -> ! List.of(preprocessed, SERVICES, HOSTS, CONFIG_DEFINITIONS_DIR).contains(name));
+        var tmpDirBase = Paths.get(System.getProperty("java.io.tmpdir", Defaults.getDefaults().underVespaHome("tmp")));
+        var tempDir = Files.createTempDirectory(tmpDirBase, "preprocessed-tempdir").toFile();
+
         File servicesFile = validateServicesFile();
-        preprocessXML(applicationFile(preprocessedDir, SERVICES), servicesFile, zone);
-        preprocessXML(applicationFile(preprocessedDir, HOSTS), getHostsFile(), zone);
+        IOUtils.copyDirectory(appDir, tempDir, -1,
+                              (__, name) -> ! List.of(preprocessed, SERVICES, HOSTS, CONFIG_DEFINITIONS_DIR).contains(name));
+        var preprocessedDir = applicationFile(appDir, preprocessed);
+        IOUtils.recursiveDeleteDir(preprocessedDir);
+        preprocessXML(applicationFile(tempDir, SERVICES), servicesFile, zone);
+        preprocessXML(applicationFile(tempDir, HOSTS), getHostsFile(), zone);
```

Comment:
```suggestion
        preprocessXML(applicationFile(tempDir, SERVICES), servicesFile, zone);
        preprocessXML(applicationFile(tempDir, HOSTS), getHostsFile(), zone);
        var preprocessedDir = applicationFile(appDir, preprocessed);
        IOUtils.recursiveDeleteDir(preprocessedDir);
```

Slightly easier to read if order is swapped here.